File: poc/candidate-b/src/smoke.ts
/** 候选 B：Crawlee HTTP 优先、Playwright 回退的阶段 1 冒烟。 */

import { mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { createHash } from 'node:crypto';
import { createRequire } from 'node:module';
import { resolve, join } from 'node:path';
import { performance } from 'node:perf_hooks';
import { CheerioCrawler, PlaywrightCrawler, type Request } from 'crawlee';
import {
  CONTROL_CLASSES,
  classifyDocument,
  extractInputPostId,
  urlSha256,
  type Classification,
} from './contract.js';

interface Attempt extends Classification {
  schema_version: '1.0';
  candidate: 'candidate-b';
  url: string;
  url_sha256: string;
  channel: 'http' | 'browser-dom';
  started_at: string;
  ended_at: string;
  duration_ms: number;
  http_status: number | null;
  final_url_sha256: string | null;
  error_category: string | null;
}

interface FinalResult {
  schema_version: '1.0';
  candidate: 'candidate-b';
  url: string;
  url_sha256: string;
  input_post_id: string;
  observed_post_id: string | null;
  post_id_matches: boolean;
  title_present: boolean;
  body_present: boolean;
  response_class: Classification['response_class'];
  control_hit: boolean;
  channel: 'http' | 'browser-dom';
  status: Classification['status'];
  request_count: number;
  started_at: string;
  ended_at: string;
  duration_ms: number;
  http_status: number | null;
  error_category: string | null;
}

interface CliOptions {
  input: string;
  outputDir: string;
  limit: number;
}

function parseArgs(argv: string[]): CliOptions {
  const values = new Map<string, string>();
  for (let index = 0; index < argv.length; index += 2) {
    const key = argv[index];
    const value = argv[index + 1];
    if (!key?.startsWith('--') || value === undefined) throw new Error(`参数格式错误: ${key ?? '<empty>'}`);
    values.set(key, value);
  }
  const input = values.get('--input');
  const outputDir = values.get('--output-dir');
  const limit = Number.parseInt(values.get('--limit') ?? '3', 10);
  if (!input || !outputDir || !Number.isInteger(limit) || limit < 1) {
    throw new Error('必须提供 --input、--output-dir，且 --limit 必须为正整数');
  }
  return { input: resolve(input), outputDir: resolve(outputDir), limit };
}

function now(): string {
  return new Date().toISOString();
}

function requestUrl(request: Request): string {
  const value = request.userData['inputUrl'];
  if (typeof value !== 'string') throw new Error('请求缺少 inputUrl');
  return value;
}

function makeAttempt(
  url: string,
  channel: 'http' | 'browser-dom',
  startedAt: string,
  started: number,
  finalUrl: string,
  httpStatus: number | null,
  document: string,
): Attempt {
  const classification = classifyDocument(finalUrl, httpStatus, document, extractInputPostId(url));
  return {
    schema_version: '1.0',
    candidate: 'candidate-b',
    url,
    url_sha256: urlSha256(url),
    channel,
    started_at: startedAt,
    ended_at: now(),
    duration_ms: Math.round(performance.now() - started),
    http_status: httpStatus,
    final_url_sha256: urlSha256(finalUrl),
    ...classification,
    error_category: classification.status === 'success' ? null : classification.response_class,
  };
}

function makeErrorAttempt(
  url: string,
  channel: 'http' | 'browser-dom',
  startedAt: string,
  started: number,
  error: unknown,
): Attempt {
  return {
    schema_version: '1.0',
    candidate: 'candidate-b',
    url,
    url_sha256: urlSha256(url),
    channel,
    started_at: startedAt,
    ended_at: now(),
    duration_ms: Math.round(performance.now() - started),
    http_status: null,
    final_url_sha256: null,
    observed_post_id: null,
    post_id_matches: false,
    title_present: false,
    body_present: false,
    response_class: 'error',
    status: 'failed',
    error_category: error instanceof Error ? error.name : 'UnknownError',
  };
}

function requestsFor(urls: string[], channel: 'http' | 'browser'): Array<{ url: string; uniqueKey: string; userData: object }> {
  return urls.map((url) => ({
    url,
    uniqueKey: `${channel}:${urlSha256(url)}`,
    userData: { inputUrl: url },
  }));
}

async function runHttp(urls: string[], outputDir: string): Promise<Map<string, Attempt>> {
  const attempts = new Map<string, Attempt>();
  const starts = new Map<string, { at: string; tick: number }>();
  const crawler = new CheerioCrawler({
    maxConcurrency: 1,
    maxRequestRetries: 0,
    retryOnBlocked: false,
    useSessionPool: false,
    requestHandlerTimeoutSecs: 45,
    preNavigationHooks: [
      async ({ request }) => {
        starts.set(requestUrl(request), { at: now(), tick: performance.now() });
      },
    ],
    requestHandler: async ({ request, response, body }) => {
      const url = requestUrl(request);
      const start = starts.get(url) ?? { at: now(), tick: performance.now() };
      const document = Buffer.isBuffer(body) ? body.toString('utf8') : String(body);
      const finalUrl = request.loadedUrl ?? url;
      const capture = join(outputDir, 'captures', `${urlSha256(url)}-http.html`);
      mkdirSync(resolve(capture, '..'), { recursive: true });
      writeFileSync(capture, document, 'utf8');
      attempts.set(url, makeAttempt(url, 'http', start.at, start.tick, finalUrl, response.statusCode ?? null, document));
    },
    failedRequestHandler: async ({ request }, error) => {
      const url = requestUrl(request);
      const start = starts.get(url) ?? { at: now(), tick: performance.now() };
      attempts.set(url, makeErrorAttempt(url, 'http', start.at, start.tick, error));
    },
  });
  await crawler.run(requestsFor(urls, 'http'));
  return attempts;
}

async function runBrowser(urls: string[], outputDir: string): Promise<Map<string, Attempt>> {
  const attempts = new Map<string, Attempt>();
  if (urls.length === 0) return attempts;
  const starts = new Map<string, { at: string; tick: number }>();
  const crawler = new PlaywrightCrawler({
    maxConcurrency: 1,
    maxRequestRetries: 0,
    retryOnBlocked: false,
    useSessionPool: false,
    requestHandlerTimeoutSecs: 60,
    launchContext: { launchOptions: { headless: true } },
    preNavigationHooks: [
      async ({ request }) => {
        starts.set(requestUrl(request), { at: now(), tick: performance.now() });
      },
    ],
    requestHandler: async ({ request, page, response }) => {
      const url = requestUrl(request);
      const start = starts.get(url) ?? { at: now(), tick: performance.now() };
      await page.waitForTimeout(5_000);
      const document = await page.content();
      const finalUrl = page.url();
      const capture = join(outputDir, 'captures', `${urlSha256(url)}-browser.html`);
      mkdirSync(resolve(capture, '..'), { recursive: true });
      writeFileSync(capture, document, 'utf8');
      attempts.set(url, makeAttempt(url, 'browser-dom', start.at, start.tick, finalUrl, response?.status() ?? null, document));
    },
    failedRequestHandler: async ({ request }, error) => {
      const url = requestUrl(request);
      const start = starts.get(url) ?? { at: now(), tick: performance.now() };
      attempts.set(url, makeErrorAttempt(url, 'browser-dom', start.at, start.tick, error));
    },
  });
  await crawler.run(requestsFor(urls, 'browser'));
  return attempts;
}

function writeJsonl(path: string, records: object[]): void {
  writeFileSync(path, `${records.map((record) => JSON.stringify(record)).join('\n')}\n`, 'utf8');
}

async function main(): Promise<void> {
  const options = parseArgs(process.argv.slice(2));
  const allUrls = readFileSync(options.input, 'utf8')
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);
  if (options.limit > allUrls.length) throw new Error('limit 超出输入清单范围');
  const urls = allUrls.slice(0, options.limit);
  mkdirSync(options.outputDir, { recursive: true });

  const overallStarts = new Map(urls.map((url) => [url, { at: now(), tick: performance.now() }]));
  const httpAttempts = await runHttp(urls, options.outputDir);
  const fallbackUrls = urls.filter((url) => httpAttempts.get(url)?.status !== 'success');
  const browserAttempts = await runBrowser(fallbackUrls, options.outputDir);
  const events: Attempt[] = [];
  const results: FinalResult[] = [];

  for (const url of urls) {
    const attempts = [httpAttempts.get(url), browserAttempts.get(url)].filter((item): item is Attempt => item !== undefined);
    if (attempts.length === 0) throw new Error(`URL 没有产生请求事件: ${urlSha256(url)}`);
    events.push(...attempts);
    const final = attempts.at(-1)!;
    const overall = overallStarts.get(url)!;
    const result: FinalResult = {
      schema_version: '1.0',
      candidate: 'candidate-b',
      url,
      url_sha256: urlSha256(url),
      input_post_id: extractInputPostId(url),
      observed_post_id: final.observed_post_id,
      post_id_matches: final.post_id_matches,
      title_present: final.title_present,
      body_present: final.body_present,
      response_class: final.response_class,
      control_hit: attempts.some((attempt) => CONTROL_CLASSES.has(attempt.response_class)),
      channel: final.channel,
      status: final.status,
      request_count: attempts.length,
      started_at: overall.at,
      ended_at: now(),
      duration_ms: Math.round(performance.now() - overall.tick),
      http_status: final.http_status,
      error_category: final.error_category,
    };
    results.push(result);
    console.log(JSON.stringify({
      url_sha256: result.url_sha256,
      status: result.status,
      response_class: result.response_class,
      channel: result.channel,
    }));
  }
  writeJsonl(join(options.outputDir, 'url-results.jsonl'), results);
  writeJsonl(join(options.outputDir, 'request-events.jsonl'), events);
  const require = createRequire(import.meta.url);
  const environment = {
    schema_version: '1.0',
    candidate: 'candidate-b',
    access_mode: 'anonymous',
    operating_system: `${process.platform}-${process.arch}`,
    node_version: process.version,
    crawlee_version: (require('crawlee/package.json') as { version: string }).version,
    playwright_version: (require('playwright/package.json') as { version: string }).version,
    input_count: urls.length,
    input_file_sha256: createHash('sha256').update(readFileSync(options.input)).digest('hex'),
  };
  writeFileSync(join(options.outputDir, 'environment.json'), `${JSON.stringify(environment, null, 2)}\n`, 'utf8');
}

await main();

File: poc/candidate-a/src/smoke.py
"""候选 A：Scrapling HTTP 优先、动态页面回退的阶段 1 冒烟。"""

from __future__ import annotations

import argparse
import hashlib
import json
import platform
import sys
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable

from scrapling.fetchers import DynamicFetcher, Fetcher
import scrapling

ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(ROOT / "poc" / "shared"))

from contract import classify_document, extract_input_post_id, url_sha256  # noqa: E402

CONTROL_CLASSES = {"rate_limited", "captcha", "challenge", "login"}


def utc_now() -> str:
    """返回带时区的 UTC ISO 8601 时间。"""

    return datetime.now(timezone.utc).isoformat()


def response_document(response: Any) -> str:
    """把 Scrapling 响应体统一转换为 UTF-8 文本。"""

    body = response.body
    if isinstance(body, bytes):
        return body.decode(getattr(response, "encoding", None) or "utf-8", errors="replace")
    return str(body)


def fetch_http(url: str) -> tuple[str, int | None, str]:
    """使用 Scrapling 普通 HTTP 通道访问单个 URL。"""

    response = Fetcher.get(url, timeout=30, retries=1)
    return str(getattr(response, "url", url)), int(response.status), response_document(response)


def fetch_dynamic(url: str) -> tuple[str, int | None, str]:
    """使用 Scrapling 标准动态页面通道访问单个 URL。"""

    response = DynamicFetcher.fetch(
        url,
        headless=True,
        wait=5,
        timeout=45_000,
        network_idle=False,
        retries=1,
    )
    return str(getattr(response, "url", url)), int(response.status), response_document(response)


def run_attempt(
    *,
    candidate: str,
    url: str,
    input_post_id: str,
    channel: str,
    fetcher: Callable[[str], tuple[str, int | None, str]],
    capture_path: Path,
) -> dict[str, Any]:
    """执行一次通道访问并返回统一请求事件。"""

    started_at = utc_now()
    started = time.perf_counter()
    try:
        final_url, http_status, document = fetcher(url)
        capture_path.parent.mkdir(parents=True, exist_ok=True)
        capture_path.write_text(document, encoding="utf-8")
        classification = classify_document(final_url, http_status, document, input_post_id)
        error_category = None if classification["status"] == "success" else classification["response_class"]
        return {
            "schema_version": "1.0",
            "candidate": candidate,
            "url": url,
            "url_sha256": url_sha256(url),
            "channel": channel,
            "started_at": started_at,
            "ended_at": utc_now(),
            "duration_ms": round((time.perf_counter() - started) * 1000),
            "http_status": http_status,
            "final_url_sha256": url_sha256(final_url),
            **classification,
            "error_category": error_category,
        }
    except Exception as exc:  # 框架异常必须形成逐 URL 证据，不能中断后静默丢失。
        return {
            "schema_version": "1.0",
            "candidate": candidate,
            "url": url,
            "url_sha256": url_sha256(url),
            "channel": channel,
            "started_at": started_at,
            "ended_at": utc_now(),
            "duration_ms": round((time.perf_counter() - started) * 1000),
            "http_status": None,
            "final_url_sha256": None,
            "observed_post_id": None,
            "post_id_matches": False,
            "title_present": False,
            "body_present": False,
            "response_class": "error",
            "status": "failed",
            "error_category": type(exc).__name__,
        }


def build_result(url: str, output_dir: Path) -> tuple[dict[str, Any], list[dict[str, Any]]]:
    """先 HTTP 后动态回退，形成一条最终结果和完整请求事件。"""

    candidate = "candidate-a"
    input_post_id = extract_input_post_id(url)
    url_id = url_sha256(url)
    started_at = utc_now()
    started = time.perf_counter()
    attempts = [
        run_attempt(
            candidate=candidate,
            url=url,
            input_post_id=input_post_id,
            channel="http",
            fetcher=fetch_http,
            capture_path=output_dir / "captures" / f"{url_id}-http.html",
        )
    ]
    if attempts[-1]["status"] != "success":
        attempts.append(
            run_attempt(
                candidate=candidate,
                url=url,
                input_post_id=input_post_id,
                channel="browser-dom",
                fetcher=fetch_dynamic,
                capture_path=output_dir / "captures" / f"{url_id}-browser.html",
            )
        )
    final = attempts[-1]
    result = {
        "schema_version": "1.0",
        "candidate": candidate,
        "url": url,
        "url_sha256": url_id,
        "input_post_id": input_post_id,
        "observed_post_id": final["observed_post_id"],
        "post_id_matches": final["post_id_matches"],
        "title_present": final["title_present"],
        "body_present": final["body_present"],
        "response_class": final["response_class"],
        "control_hit": any(attempt["response_class"] in CONTROL_CLASSES for attempt in attempts),
        "channel": final["channel"],
        "status": final["status"],
        "request_count": len(attempts),
        "started_at": started_at,
        "ended_at": utc_now(),
        "duration_ms": round((time.perf_counter() - started) * 1000),
        "http_status": final["http_status"],
        "error_category": final["error_category"],
    }
    return result, attempts


def write_jsonl(path: Path, records: list[dict[str, Any]]) -> None:
    """以 UTF-8/LF 写出 JSONL。"""

    payload = "".join(json.dumps(record, ensure_ascii=False, separators=(",", ":")) + "\n" for record in records)
    path.write_bytes(payload.encode("utf-8"))


def write_environment(path: Path, input_path: Path, input_count: int) -> None:
    """记录阶段 1 冒烟的最小可复现环境，不采集凭证或无关机器文件。"""

    payload = {
        "schema_version": "1.0",
        "candidate": "candidate-a",
        "access_mode": "anonymous",
        "operating_system": platform.platform(),
        "python_version": platform.python_version(),
        "python_executable": str(Path(sys.executable).resolve()),
        "scrapling_version": scrapling.__version__,
        "input_count": input_count,
        "input_file_sha256": hashlib.sha256(input_path.read_bytes()).hexdigest(),
    }
    path.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", type=Path, required=True)
    parser.add_argument("--output-dir", type=Path, required=True)
    parser.add_argument("--limit", type=int, default=3)
    args = parser.parse_args()

    urls = [line.strip() for line in args.input.read_text(encoding="utf-8-sig").splitlines() if line.strip()]
    if args.limit < 1 or args.limit > len(urls):
        raise ValueError("limit 超出输入清单范围")
    urls = urls[: args.limit]
    args.output_dir.mkdir(parents=True, exist_ok=True)
    results: list[dict[str, Any]] = []
    events: list[dict[str, Any]] = []
    for url in urls:
        result, attempts = build_result(url, args.output_dir)
        results.append(result)
        events.extend(attempts)
        print(
            json.dumps(
                {
                    "url_sha256": result["url_sha256"],
                    "status": result["status"],
                    "response_class": result["response_class"],
                    "channel": result["channel"],
                },
                ensure_ascii=False,
            )
        )
    write_jsonl(args.output_dir / "url-results.jsonl", results)
    write_jsonl(args.output_dir / "request-events.jsonl", events)
    write_environment(args.output_dir / "environment.json", args.input, len(urls))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
